### DISCUSSION TITLE
sqlalchemy.exc.InvalidRequestError: Class does not have a __table__ or __tablename__ specified and does not inherit from an existing table-mapped class

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#8699] [Usage Questions] [answered]
I'm trying to use mixin and MappedAsDataclass in 2.0.0b1, and make them easy to extend, including using them with alembic

example:

```python
from datetime import datetime
from typing import Optional

from sqlalchemy import func
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, declared_attr, MappedAsDataclass
from typing_extensions import Annotated

# Universal primary key
id_key = Annotated[int, mapped_column(init=False, primary_key=True, index=True, autoincrement=True)]


class _BaseMixin(MappedAsDataclass):

    @declared_attr.directive
    def __tablename__(cls) -> Optional[str]:
        return cls.__name__.lower()

    id: Mapped[id_key]
    create_user: Mapped[int]
    update_user: Mapped[Optional[int]]
    created_time: Mapped[datetime] = mapped_column(init=False, default=func.now())
    updated_time: Mapped[Optional[datetime]] = mapped_column(init=False, onupdate=func.now())

    __abstract__ = True


class MappedBase(DeclarativeBase):
    ...


class DataClassBase(MappedAsDataclass, MappedBase):
    ...


class Base(_BaseMixin, MappedBase):
    ...
```
For people who don't routinely apply dataclass, they use `MappedBase`, instead, they use `DataClassBase` or `Base`.

In the models's \_\_init__.py and alembic env file, use `MappedBase`, in my opinion.

alembic env:

```python
target_metadata = MappedBase.metadata
```

models package:

```python
from backend.app.database.base_class import MappedBase
from backend.app.models.user import User
from backend.app.models.t import TB
```

User model:

```python

class User(DataClassBase):
    """ user table """
    __tablename__ = 'sys_user'

    id: Mapped[id_key]
    username: Mapped[str] = mapped_column(String(20), unique=True, index=True)
    password: Mapped[str] = mapped_column(String(255))
    email: Mapped[str] = mapped_column(String(128), unique=True, index=True)
    is_superuser: Mapped[bool] = mapped_column(default=False)
    is_active: Mapped[bool] = mapped_column(default=True)
    time_joined: Mapped[datetime] = mapped_column(init=False, default=func.now())
    last_login: Mapped[Optional[datetime]] = mapped_column(init=False, default=None, onupdate=func.now())
```

TB model:

```python
class TB(Base):
    """ test Base """
    test: Mapped[str] = mapped_column(String(10))
```

shell exec:

```shell
alembic revision --autogenerate
```

sqlalchemy.exc.InvalidRequestError:

```txt
sqlalchemy.exc.InvalidRequestError: Class <class 'backend.app.database.base_class.DataClassBase'> does not have a __table__ or __tablename__ specified and does not inherit from an existing table-mapped class.
```

**Comment by wu-clan** [comment]
I don't seem to see any cases in the documentation where `Mixin`, `MappedAsDataclass`, `DeclarativeBase` are used together

**Comment by CaselIT** [comment]
I believe you forgot to specify `__abstract__=True`

**Comment by wu-clan** [comment]
Yes, adding `__abstract__=True` to `DataClassBase` worked, but I can't understand the reasoning behind this, I think the left and right possibility is to use `__abstract__=True` in `Base`, but not

**Reply by zzzeek** [reply]
any class that extends from DeclarativeBase must:

1. be mapped - to be mapped, a class must:
    a. have a `__tablename__` or `__table__` attribute, so we know what table it's mapped to, or
    b. extend from another class that has a `__tablename__` or `__table__` attribute, in which case we set it up as "single table inheritance"
    or
2. not be mapped, which is indicated by `__abstract__ = True`
3. a class that doesnt have the `__tablename__` / `__table__` req above and also is not `__abstract__`, we would have to make an assumption, which is awkward because a class without `__table(name)__` or `__abstract__` is still valid in the single inheritance case.  so, we refuse the temptation to guess and raise an error.

**Comment by wu-clan** [comment, marked as answer]
Thanks a lot, I think I understand, but is it stated somewhere in the documentation? Forgive my memory, I think it's necessary to add a note, although in most cases `__tablename__` is used, but multiple inheritance like this does use `__abstract__=True`

**Reply by zzzeek** [reply]
`__abstract__` is mentioned here: https://docs.sqlalchemy.org/en/14/orm/declarative_config.html#abstract 

I *thought* it was mentioned at https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html but it isn't, which is bad!  it definitely should be.

**Reply by wu-clan** [reply]
Thanks again, maybe it should be added later, I think I can tick this answer
